fx/nnc_compile.py

```diff
@@ -0,0 +1,448 @@
+
+def transpose_lower(name, out_shape, inp_shapes, args):
+    idx_1, idx_2 = args[1], args[2]
+    def transpose(shape):
+        shape[idx_1], shape[idx_2] = shape[idx_2], shape[idx_1]
+        return shape
+    def f(*idxs):
+        idxs = transpose(list(idxs))
+        return args[0].load(idxs)
+    return te.Compute(name, get_dim_args(out_shape), f)
+
+def flatten_lower(name, out_shape, inp_shapes, args):
+    A, start_dim, end_dim = args
+    shape = list(inp_shapes[0][0])
+    flattened_region = shape[start_dim:end_dim+1]
+    def prod(x):
+        t = 1
+        for i in x:
+            t *= i
+        return t
+    def get_orig_idxs(i):
+        idxs = []
+        total = prod(flattened_region)
+        for dim in flattened_region:
+            total //= dim
+            idxs.append(i / to_expr(total))
+            i = i % to_expr(total)
+        return idxs
+    def f(*idxs):
+        idxs = list(idxs)
+        idxs = idxs[:start_dim] + get_orig_idxs(idxs[start_dim]) + idxs[start_dim+1:]
+        return A.load(idxs)
+    return te.Compute(name, get_dim_args(out_shape), f)
+
+def cat_lower(name, out_shape, inp_shapes, args):
+    tensors = args[0]
+    dim = args[1]
+    lengths = [i[0][dim] for i in inp_shapes[0]]
+    def f(*idxs):
+        idxs = list(idxs)
+        sm = lengths[0]
+        load = tensors[0].load(idxs)
+        for length, tensor in list(zip(lengths, tensors))[1:]:
+            new_idxs = idxs[:]
+            new_idxs[dim] -= to_expr(sm)
+            load = te.ifThenElse(idxs[dim] < to_expr(sm), load, tensor.load(new_idxs))
+        return load
+    return te.Compute(name, get_dim_args(out_shape), f)
+
+lowering_functions[torch.transpose] = transpose_lower
+lowering_functions[torch.flatten] = flatten_lower
+lowering_functions[torch.cat] = cat_lower
+
+def bmm_lower(name, out_shape, inp_shapes, args):
+    M1 = args[0]
+    M2 = args[1]
+    B, N, M = inp_shapes[0][0]
+    P = inp_shapes[1][0][2]
+
+    def f(b, n, p, m):
+        return M1.load([b, n, m]) * M2.load([b, m, p])
+    mm = te.Compute('mm', get_dim_args([B,N,P,M]), f)
+    return te.SumReduce(name, get_dim_args([B, N, P]), mm, get_dim_args([M]))
+
+
+def mm_lower(name, out_shape, inp_shapes, args):
+    M1 = args[0]
+    M2 = args[1]
+    N, M = inp_shapes[0][0]
+    P = inp_shapes[1][0][1]
+
+    def f(n, p, m):
+        return M1.load([n, m]) * M2.load([m, p])
+    mm = te.Compute('mm', get_dim_args([N,P,M]), f)
+    return te.SumReduce(name, get_dim_args([N, P]), mm, get_dim_args([M]))
+
+lowering_functions[torch.bmm] = bmm_lower
+lowering_functions[torch.mm] = mm_lower
+
+
+def lower_function(node, op, nnc_args, args):
+    inp_shapes = fx.node.map_aggregate(args, lambda arg: (arg.shape, arg.dtype) if isinstance(arg, fx.Node) else None)
+    return lowering_functions[op](node.name, node.shape, inp_shapes, nnc_args)
+
+def nnc_compile(model: torch.nn.Module, example_inputs) -> torch.nn.Module:
```

Comment:
Probably want to add a docstring and comments within this function to describe the high-level process